external/fv3fit/fv3fit/data/synthetic.py

```diff
@@ -0,0 +1,238 @@
+        return dacite.from_dict(
+            data_class=cls, data=d, config=dacite.Config(strict=True)
+        )
+
+
+@register_tfdataset_loader
+@dataclasses.dataclass
+class SyntheticWaves(TFDatasetLoader):
+    """
+    Attributes:
+        nsamples: number of samples to generate per batch
+        nbatch: number of batches to generate
+        nx: length of x- and y-dimensions to generate
+        nz: length of z-dimension to generate
+        scalar_names: names to generate as scalars instead of
+            vertically-resolved variables
+        scale_min: minimum amplitude of waves
+        scale_max: maximum amplitude of waves
+        period_min: minimum period of waves
+        period_max: maximum period of waves
+        phase_range: fraction of 2*pi to use for possible range of
+            random phase, should be a value between 0 and 1.
+        wave_type: one of "sinusoidal" or "square"
+    """
+
+    nsamples: int
+    nbatch: int
+    ntime: int
+    nx: int
+    nz: int
+    wave_type: str
+    scalar_names: List[str] = dataclasses.field(default_factory=list)
+    scale_min: float = 0.0
+    scale_max: float = 1.0
+    period_min: float = 8.0
+    period_max: float = 16.0
+    phase_range: float = 1.0
+
+    def open_tfdataset(
+        self, local_download_path: Optional[str], variable_names: Sequence[str],
+    ) -> tf.data.Dataset:
+        """
+        Args:
+            local_download_path: if provided, cache data locally at this path
+            variable_names: names of variables to include when loading data
+        Returns:
+            dataset containing requested variables, each record is a mapping from
+                variable name to variable value, and each value is a tensor whose
+                first dimension is the batch dimension
+        """
+        if self.wave_type == "sinusoidal":
+            func = np.sin
+        elif self.wave_type == "square":
+
+            def func(x):
+                return np.sign(np.sin(x))
+
+        else:
+            raise ValueError(f"Invalid wave_type {self.wave_type}")
+
+        dataset = get_waves_tfdataset(
+            variable_names,
+            scalar_names=self.scalar_names,
+            nsamples=self.nsamples,
+            nbatch=self.nbatch,
+            ntime=self.ntime,
+            nx=self.nx,
+            ny=self.nx,
+            nz=self.nz,
+            scale_min=self.scale_min,
+            scale_max=self.scale_max,
+            period_min=self.period_min,
+            period_max=self.period_max,
+            phase_range=self.phase_range,
+            func=func,
+        )
+        if local_download_path is not None:
+            dataset = dataset.cache(local_download_path)
+        return dataset
+
+    @classmethod
+    def from_dict(cls, d: dict) -> "TFDatasetLoader":
+        return dacite.from_dict(
+            data_class=cls, data=d, config=dacite.Config(strict=True)
+        )
+
+
+def get_waves_tfdataset(
+    variable_names,
+    *,
```

Comment:
Just curious what's the reason for needing only the variables names list as a positional?

Reply:
Personal preference honestly. Whenever I call a function with many arguments, whether it's required to or not I try to put the first argument as positional and the rest of the arguments as keyword. Usually the first argument is obvious enough from the function and feels like the "object" while the other arguments feel like "descriptors" or settings (which I'd say is true here).